### PR TITLE
#997 Lookup table type inference

### LINKED ISSUE #997: Lookup tables: Parse JSON strings to objects on import
In order to be able to imported the exported Medsafe data to a Lookup table, the import module will need to be able to parse JSON strings (from the CSV file) and save them as `jsonb` objects in the database. Currently, it treats everything as a string.

### PR DESCRIPTION
**Pull request by CarlosNZ**
Fix #997 

Requires front-end https://github.com/openmsupply/conforma-web-app/pull/1471 in order for the UI to work with this.

Tried to be as minimally invasive to the existing code as possible, so this just intercepts the existing imported data and mutates the fieldMap and data rows in-place to conform to inferred data type. The postgres fields will be one of:
- `varchar` (what they all are by default)
- `integer`
- `double precision`
- `boolean`
- `timestamptz`
- `jsonb` (the one we really want for scraped data import)

Test it with the following.

Import this lookup table, and notice the different data types in the created table:
[test_data.csv](https://github.com/openmsupply/conforma-server/files/10424847/test_data.csv)

Then import this one over the top to see the effect of modifying the exported CSV.
[test_2023-01-16.csv](https://github.com/openmsupply/conforma-server/files/10424848/test_2023-01-16.csv)

You can export it again yourself and change/add some data to check re-import as well.

### CODE REVIEW

File: src/lookup-table/models/LookupTable.model.ts [hunk 1]
```diff
@@ -158,7 +159,7 @@ const LookupTableModel = () => {
     tableName: string,
     fieldMaps: FieldMapType[]
   ): Promise<boolean> => {
-    const text = `UPDATE data_table SET field_map = $1 WHERE display_name = $2`
+    const text = `UPDATE data_table SET field_map = $1 WHERE table_name = $2`
```
Comment (CarlosNZ) [inline src/lookup-table/models/LookupTable.model.ts:162]: This was an existing mistake -- the field maps weren't getting updated when new columns were added due to this.

File: src/lookup-table/utils/dataTypeUtils.ts [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+        const type = getType(value)
+        if (type !== null) {
+          fieldMaps[index].dataType = type
+          done.push(key)
+        }
+      }
+    })
+
+    if (done.length === fieldMaps.length) break
+  }
```
Comment (CarlosNZ) [inline src/lookup-table/utils/dataTypeUtils.ts:20]: It just checks the first value of each column to determine its type. However, we can have `null` values, so it iterates through all rows until it's found at least one non-null value for each field, then it breaks out of the loop.

File: src/lookup-table/utils/dataTypeUtils.ts [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+
+  const fieldMap: { [key: string]: any } = fieldMaps.reduce((acc, curr) => {
+    const name = curr.fieldname
+    const type = curr.dataType
+    return { ...acc, [name]: type }
+  }, {})
```
Comment (CarlosNZ) [inline src/lookup-table/utils/dataTypeUtils.ts:26]: This is just a compact version of the main `fieldMaps` array, with a simple `{ key: type }` object. Makes the subsequent lookup a bit cleaner.

File: src/lookup-table/utils/dataTypeUtils.ts [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+
+  // Now mutate rows based on those types
+  rows.forEach((row, index) => {
+    rows[index] = mapValues(row, (value, key) => {
+      if (key === 'id') return value
+      return convertType(value, fieldMap[key])
```
Comment (CarlosNZ) [inline src/lookup-table/utils/dataTypeUtils.ts:32]: All values coming from CSV files are strings, so this converts the actual data to the correct type (i.e. `"2"` becomes `2`, `"FALSE"` becomes `false` etc.

File: src/lookup-table/utils/dataTypeUtils.ts [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+// Prepares table export by stringifying jsonb fields
+// Mutates rows *in-place*
+export const exportDataRows = (fieldMaps: FieldMapType[], rows: { [key: string]: any }[]) => {
+  const jsonFields = fieldMaps.filter((e) => e.dataType === 'jsonb').map((e) => e.fieldname)
+
+  rows.forEach((row) => {
+    jsonFields.forEach((jsonField: string) => {
+      row[jsonField] = row[jsonField] !== null ? JSON.stringify(row[jsonField]) : ''
+    })
+  })
```
Comment (CarlosNZ) [inline src/lookup-table/utils/dataTypeUtils.ts:46]: Without this, the json fields get exported as `[Object object]`, so this stringifies those fields. All other fields get dumped out as strings (like they were when imported).

**Review by nmadruga — APPROVED** [review]
Working well
Code looks good

**Pull request merged**